### PR TITLE
Mixer scheduler

### PR DESCRIPTION
**Pull request by raphaelcoeffic**
Re-work of the mixer scheduling. It uses a separate timer (TIM13) to trigger a CoOS flag, which is waited for in the mixer loop.

TODO:
- [x] fix/test on X9D(+)
- [ ] deal with `SKY9X`

### CODE REVIEW

**Comment by ctzsnooze** [comment]
Hi Raphael - this is great work - any chance for a QX7 binary for testing?

**Comment by JyeSmith** [comment]
Hey @raphaelcoeffic I'm also keen to try it out when you're ready for testers (QX7).

**Comment by raphaelcoeffic** [comment]
@ctzsnooze @JyeSmith Guys, I just rebased on top of current 2.3 branch. Seems I still have to fix the lag/refresh rate display for crossfire. After that we can have a test version, but please take into account that this feature has been refused for 2.3 and will be integrated first into 2.4.

**Comment by AlessandroAU** [comment]
Keen to get this working in order to improve latency!

**Comment by JyeSmith** [comment]
@raphaelcoeffic what is the best way implement the heartbeat on a module?  Say you want to request packets at 200Hz?

How exactly does it work 😅

**Comment by raphaelcoeffic** [comment]
@JyeSmith the module needs to report two things: the requested period (and thus frequency), and the current lag.

After that, the frequency is adapted to reduce lag and be equal the requested frequency.
The telemetry packets sent by multimodule and crossfire can be used as an example (crossfire is probably best as it contains only these data).

**Comment by AlessandroAU** [comment]
What is the current lag? The time difference between when the TX module wants to send the packet and when it is received from opentx?

**Comment by raphaelcoeffic** [comment]
@AlessandroAU for you module, as it implements the crossfire protocol, all you need is to send the proper telemetry packets (should be easy to find in this MR). Let me know if you need any help.

**Comment by raphaelcoeffic** [comment]
> What is the current lag? The time difference between when the TX module wants to send the packet and when it is received from opentx?

Yes, exactly.

**Comment by AlessandroAU** [comment]
I'm a bit draft and still can't seem to figure out it,

can you give me an example of what the tlm packet would look like?

Edit: This format for the 'sync packet' seems to work 
image

**Comment by AlessandroAU** [comment]
@raphaelcoeffic Any recommendations of how often the sync update packet should be sent to the TX?

File: radio/src/mixer_scheduler.h [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+ * GNU General Public License for more details.
+ */
+
+#ifndef _MIXER_SCHEDULER_H_
+#define _MIXER_SCHEDULER_H_
+
+#define MIXER_SCHEDULER_DEFAULT_PERIOD_US 4000u // 4ms
+
+#define MIN_REFRESH_RATE      4000 /* us */
+#define MAX_REFRESH_RATE     25000 /* us */
```
Comment (JyeSmith) [inline radio/src/mixer_scheduler.h:27]: Is it possible to have a longer MAX_REFRESH_RATE?  ExLRS goes as low as 25Hz and I believe Ghost goes down to 15Hz.

**Comment by robertlacroix** [comment]
@raphaelcoeffic anything we can help with to get this mergeable?

**Comment by 3djc** [comment]
Outdated, so closing it. Merge is currently planed for 2.4

**Pull request closed without merging**